security/elastic.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2016: Mirantis Inc.
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import json
+import logging
+
+import elasticsearch
+from elasticsearch import helpers
+
+from security import base
+
+INDEX = "ms_security_"
+
+
+class StorageException(Exception):
+    pass
+
+
+class Backend(object):
+    """Represent elasticsearch backend.
+
+    usage:
+        b = elastic.Backend(urls=["url1", "url2"])
+        issues = b.get_issues("Region1", "IssueType1,IssueType2")
+        b.begin()
+        b.store_issue(Issue(...))
+        b.commit("region1")
+    """
+
+    def __init__(self, urls, **kwargs):
+        self.index = INDEX
+        self.es = elasticsearch.Elasticsearch(urls)
+        self.body = ""
+
+    def get_issues(self, region, types=None, discovered_after=None):
+        query = {
+            "query": {
+                "bool": {
+                    "must_not": {
+                        "exists": {
+                            "field": "resolvedAt",
+                        },
+                    },
+                },
+            }
+        }
+        if discovered_after:
+            query["query"]["bool"]["must"] = {
+                "range": {
+                    "discoveredAt": {"gt": discovered_after.isoformat("T")}
```

Comment:
this is not optimal in ElasticSearch, you should always round datetime https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#_date_math_and_rounding otherwise Elasticsearch won't cache this filter